core/types/receipt_test.go

```diff
@@ -745,6 +810,46 @@ func TestBedrockDepositReceiptUnchanged(t *testing.T) {
 	require.EqualValues(t, receipt.Logs, parsed.Logs)
 	// And still shouldn't have a nonce
 	require.Nil(t, parsed.DepositNonce)
+	// ..or a deposit nonce
+	require.Nil(t, parsed.DepositReceiptVersion)
+}
+
+// Regolith introduced an inconsistency in behavior between EncodeIndex and MarshalBinary for a
+// deposit transaction receipt. TestReceiptEncodeIndexBugIsEnshrined makes sure this difference is
+// preserved for backwards compatibility purposes, but also that there is no discrepancy for the
+// post-Canyon encoding.
+func TestReceiptEncodeIndexBugIsEnshrined(t *testing.T) {
+	// Check that a post-Regolith, pre-Canyon receipt produces the expected difference between
+	// EncodeIndex and MarshalBinary.
+	buf := new(bytes.Buffer)
+	receipts := Receipts{depositReceiptWithNonce}
+	receipts.EncodeIndex(0, buf)
+	indexBytes := buf.Bytes()
+
+	regularBytes, _ := receipts[0].MarshalBinary()
+
+	require.NotEqual(t, indexBytes, regularBytes)
+
+	// Confirm the buggy encoding is as expected, which means it should encode as if it had no
+	// nonce specified (like that of a non-deposit receipt, whose encoding would differ only in the
+	// type byte).
+	buf.Reset()
+	tempReceipt := *depositReceiptWithNonce
+	tempReceipt.Type = eip1559Receipt.Type
+	buggyBytes, _ := tempReceipt.MarshalBinary()
+
+	require.Equal(t, indexBytes[1:], buggyBytes[1:])
+
+	// check that the post-Canyon encoding has no differences between EncodeIndex and
+	// MarshalBinary.
+	buf.Reset()
+	receipts = Receipts{depositReceiptWithNonceAndVersion}
+	receipts.EncodeIndex(0, buf)
+	indexBytes = buf.Bytes()
+
+	regularBytes, _ = receipts[0].MarshalBinary()
+
+	require.Equal(t, indexBytes, regularBytes)
```

Comment:
Would it also be worth testing that if we modify only the `DepositNonce` that we get a different output from `EncodeIndex` when a version is present?

Reply:
good idea, will add.

Reply:
done.